src/main/scala/chisel3/util/experimental/decode/EspressoMinimizer.scala

```diff
@@ -0,0 +1,75 @@
+// SPDX-License-Identifier: Apache-2.0
+
+package chisel3.util.experimental.decode
+
+import chisel3.util.BitPat
+import logger.LazyLogging
+
+case object EspressoNotFoundException extends Exception
+
+object EspressoMinimizer extends Minimizer with LazyLogging {
+  def minimize(table: TruthTable): TruthTable =
+    TruthTable.merge(TruthTable.split(table).map{case (table, indexes) => (espresso(table), indexes)})
+
+  def espresso(table: TruthTable): TruthTable = {
+    def writeTable(table: TruthTable): String = {
+      def invert(string: String) = string
+        .replace('0', 't')
+        .replace('1', '0')
+        .replace('t', '1')
+      val defaultType: Char = {
+        val t = table.default.toString.drop(7).dropRight(1).toCharArray.distinct
+        require(t.length == 1, "Internal Error: espresso only accept unified default type.")
+        t.head
+      }
+      val tableType: String = defaultType match {
+        case '?' => "fr"
+        case _ => "fd"
+      }
+      val rawTable = table
+        .toString
+        .split("\n")
+        .filter(_.contains("->"))
+        .mkString("\n")
+        .replace("->", " ")
+        .replace('?', '-')
+      // invert all output, since espresso cannot handle default is on.
+      val invertRawTable = rawTable
+        .split("\n")
+        .map(_.split(" "))
+        .map(row => s"${row(0)} ${invert(row(1))}")
+        .mkString("\n")
+      s""".i ${table.inputWidth}
+         |.o ${table.outputWidth}
+         |.type ${tableType}
+         |""".stripMargin ++ (if (defaultType == '1') invertRawTable else rawTable)
+    }
+
+    def readTable(espressoTable: String): Map[BitPat, BitPat] = {
+      def bitPat(espresso: String): BitPat = BitPat("b" + espresso.replace('-', '?'))
+
+      espressoTable
+        .split("\n")
+        .filterNot(_.startsWith("."))
+        .map(_.split(' '))
+        .map(row => bitPat(row(0)) -> bitPat(row(1)))
+        .toMap
+    }
+
+    // Since Espresso don't implements pipe, we use a temp file to do so.
+    val input = writeTable(table)
+    logger.trace(s"""espresso input table:
+                    |$input
+                    |""".stripMargin)
+    val f = os.temp(input)
+    val o = try {
+      os.proc("espresso", f).call().out.chunks.mkString
+    } catch {
+      case _ : java.io.IOException => throw EspressoNotFoundException
+    }
```

Comment:
This is better but still just fundamentally catching an `IOException`, what if the user lacks permissions or something else? It would be helpful to distinguish between possible error modes. I'm fine with only throwing `EspressoNotFoundException` if the exit code is 127 and just forwarding any other exception (eg. lacking permissions) because we don't want to mislead users as to why it's failing.

Reply:
Oh, 
```
> try {os.proc("non_exist_program").call()} catch { case e:java.io.IOException =>println(e.isInstanceOf[java.io.FileNotFoundException])}

false
```

Reply:
```
> try {os.proc("/").call()} catch { case e: java.io.IOException if e.getMessage.contains("error=2, No such file or directory") =>println("catched")}

java.io.IOException: Cannot run program "/" (in directory "xxx"): error=13, Permission denied
  java.lang.ProcessBuilder.start(ProcessBuilder.java:1141)
  java.lang.ProcessBuilder.start(ProcessBuilder.java:1072)
  os.proc.proc$lzycompute$1(ProcessOps.scala:128)
  os.proc.proc$1(ProcessOps.scala:122)
  os.proc.spawn(ProcessOps.scala:135)
  os.proc.call(ProcessOps.scala:76)
  ammonite.$sess.cmd24$.liftedTree1$1(cmd24.sc:111)
  ammonite.$sess.cmd24$.<clinit>(cmd24.sc:111)
java.io.IOException: error=13, Permission denied
  java.lang.ProcessImpl.forkAndExec(Native Method)
  java.lang.ProcessImpl.<init>(ProcessImpl.java:313)
  java.lang.ProcessImpl.start(ProcessImpl.java:244)
  java.lang.ProcessBuilder.start(ProcessBuilder.java:1108)
  java.lang.ProcessBuilder.start(ProcessBuilder.java:1072)
  os.proc.proc$lzycompute$1(ProcessOps.scala:128)
  os.proc.proc$1(ProcessOps.scala:122)
  os.proc.spawn(ProcessOps.scala:135)
  os.proc.call(ProcessOps.scala:76)
  ammonite.$sess.cmd24$.liftedTree1$1(cmd24.sc:111)
  ammonite.$sess.cmd24$.<clinit>(cmd24.sc:111)


> try {os.proc("a").call()} catch { case e: java.io.IOException if e.getMessage.contains("error=2, No such file or directory") =>println("catched")}
catched
```
ugly but I think it works.

Reply:
Fixed in cd1b9df

Reply:
What I was really thinking is `.call(check = false)` so that we could just check the error code directly, but that doesn't seem to work at the moment (see https://github.com/com-lihaoyi/os-lib/issues/82) so I guess this is fine.